### PR TITLE
Added Seq2Seq tasks

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Add Translation/Summarization and general Seq2Seq Task. This is a first pass, and will need more iteration in the future:

Painpoints:
- I didn't have enough time to train a mt5 model on translation, as this would be common for all translation tasks. This is due to requiring some additional pre-processing, and longer training times. As a stop gap, I've used a student model trained on the en/ro translation task as I think is the case in HF, but that shouldn't be what we deliver in the future.
- Upload WMT translation sets in different languages!

Left to do:

- [x] Train Summarization model, ensure this works with predict
- [x] Clean up Translation predict output, check single string input works
- [x] Upload datasets/models to S3
- [x] Create tests replicating classification tests
- [x] In a separate PR (done by someone familiar with the doc structure if poss) we need to add these additional tasks

### CODE REVIEW

**Review by carmocca — APPROVED** [review]

File: docs/source/reference/summarization.rst [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+	# import our libraries
+	from flash.text import SummarizationTask
+
+
+    # 2. Load the model from a checkpoint
+    model = SummarizationTask.load_from_checkpoint("https://flash-weights.s3.amazonaws.com/summarization_model_xsum.pt")
+
+    # 2. Perform inference from a sequence
+    predictions = model.predict([
```
Comment (Borda) [inline docs/source/reference/summarization.rst:34]: formatting

File: docs/source/reference/summarization.rst [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        """
+        Camilla bought a box of mangoes with a Brixton Â£10 note, introduced last year to try to keep the money of local
+        people within the community.The couple were surrounded by shoppers as they walked along Electric Avenue.
+        They came to Brixton to see work which has started to revitalise the borough.
+        It was Charles' first visit to the area since 1996, when he was accompanied by the former
+        South African president Nelson Mandela.Greengrocer Derek Chong, who has run a stall on Electric Avenue
+        for 20 years, said Camilla had been ""nice and pleasant"" when she purchased the fruit.
```
Comment (Borda) [inline docs/source/reference/summarization.rst:41]: curious what `""` means?

File: docs/source/reference/summarization.rst [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    print(predictions)
+
+Or on a given dataset:
+
+.. code-block:: python
+
+	# import our libraries
+	from pytorch_lightning import Trainer
+	from flash import download_data
+	from flash.text import SummarizationData, SummarizationTask
```
Comment (Borda) [inline docs/source/reference/summarization.rst:65]: formatting

File: docs/source/reference/summarization.rst [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+-----
+
+**********
+Finetuning
+**********
+
+Say you want to finetune to your own summarization data. We use the XSUM dataset as an example which contains a ``train.csv`` and ``valid.csv``, structured like so:
+
+.. code-block::
```
Comment (Borda) [inline docs/source/reference/summarization.rst:90]: how this is codeblock?

File: docs/source/reference/translation.rst [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+In the above the input/target columns represent the English and Romanian translation respectively.
+
+All we need is three lines of code to train our model!
+
+.. code-block:: python
+
+	# import our libraries
+	import flash
+	from flash import download_data
+	from flash.text import TranslationData, TranslationTask
```
Comment (Borda) [inline docs/source/reference/translation.rst:88]: this formatting issue seems to be everywhere... maybe tab vs spaces

**Review by Borda — CHANGES_REQUESTED** [review]

File: flash/text/seq2seq/core/data.py [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+from datasets import load_dataset
+from datasets.utils.download_manager import GenerateMode
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from torch import Tensor
+from transformers import AutoTokenizer, default_data_collator
+
+from flash.core.data import DataModule, TaskDataPipeline
+
+
+def prepare_dataset(
```
Comment (Borda) [inline flash/text/seq2seq/core/data.py:26]: shall this be part (static) [art of the data-pipeline?

File: flash/text/seq2seq/core/data.py [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+    def before_collate(self, samples: Any) -> Any:
+        """Override to apply transformations to samples"""
+        if isinstance(samples, (list, tuple)) and len(samples) > 0 and all(isinstance(s, str) for s in samples):
+            return [self._tokenize_fn({self._input: s, self._target: None}) for s in samples]
+        return samples
+
+    @staticmethod
+    def _tokenize_fn(
+        ex, tokenizer, input: str, target: Optional[str], max_source_length: int, max_target_length: int,
+        padding: Union[str, bool]
```
Comment (Borda) [inline flash/text/seq2seq/core/data.py:110]:
```suggestion
        padding: Union[str, bool],
```

File: flash/text/seq2seq/core/model.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+        optimizer: Type[torch.optim.Optimizer] = torch.optim.Adam,
+        metrics: Union[pl.metrics.Metric, Mapping, Sequence, None] = None,
+        learning_rate: float = 5e-5,
+        val_target_max_length: Optional[int] = None,
+        num_beams: Optional[int] = None,
+        freeze_embeds: bool = True
+    ):
+        os.environ["TOKENIZERS_PARALLELISM"] = "TRUE"
+        # disable HF thousand warnings
+        warnings.simplefilter("ignore")
```
Comment (Borda) [inline flash/text/seq2seq/core/model.py:64]: shall it be back enabled on object destroy, in `__del__`

File: flash/text/seq2seq/translation/metric.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+
+
+def _count_ngram(ngram_input_list: List[str], n_gram: int) -> Counter:
+    """
+    Counting how many times each word appears in a given text with ngram
+    Args:
+        ngram_input_list: A list of translated text or reference texts
+        n_gram: gram value ranged 1 to 4
+    Return:
+        ngram_counter: a collections.Counter object of ngram
```
Comment (Borda) [inline flash/text/seq2seq/translation/metric.py:33]:
```suggestion

    Args:
        ngram_input_list: A list of translated text or reference texts
        n_gram: gram value ranged 1 to 4

    Return:
        ngram_counter: a collections.Counter object of ngram
```

File: flash/text/seq2seq/translation/metric.py [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+            ngram_key = tuple(ngram_input_list[j:(i + j)])
+            ngram_counter[ngram_key] += 1
+
+    return ngram_counter
+
+
+class BLEUScore(Metric):
+    """
+    Calculate BLEU score of machine translated text with one or more references.
+    Example:
```
Comment (Borda) [inline flash/text/seq2seq/translation/metric.py:49]:
```suggestion

    Example:
```

File: flash_examples/predict/translate.py [hunk 11]
```diff
@@ -0,0 +1,27 @@
+from pytorch_lightning import Trainer
+
+from flash import download_data
+from flash.text import TranslationData, TranslationTask
+
+if __name__ == "__main__":
```
Comment (Borda) [inline flash_examples/predict/translate.py:6]: what is the meaning of this, if all is ignored anyway drop it from testing...

File: requirements.txt [hunk 12]
```diff
@@ -8,4 +8,6 @@ datasets==1.2.1
 pandas==1.1.2
 scikit-learn==0.24.0
 numpy
-tqdm
\ No newline at end of file
+tqdm
+rouge-score
+sentencepiece
```
Comment (Borda) [inline requirements.txt:13]: some min version?

File: docs/source/reference/summarization.rst (hunk 2 above)
Comment (SeanNaren) [reply docs/source/reference/summarization.rst:41]: I think its the way that quotations are handled via the CSV. this came from the `pandas.to_csv` function

File: flash/text/seq2seq/core/data.py (hunk 6 above)
Comment (SeanNaren) [reply flash/text/seq2seq/core/data.py:26]: Eventually it should be! but I've mimicked the text classification pipeline. I think eventually this will be consolidated in time

File: docs/source/reference/summarization.rst (hunk 4 above)
Comment (SeanNaren) [reply docs/source/reference/summarization.rst:90]: i copied this from the `text_classification.rst`, let me know what it should be

File: docs/source/reference/summarization.rst (hunk 3 above)
Comment (SeanNaren) [reply docs/source/reference/summarization.rst:65]: What should the format be? Again copied from the `text_classification.rst`

File: flash_examples/predict/translate.py (hunk 11 above)
Comment (SeanNaren) [reply flash_examples/predict/translate.py:6]: Not sure if I understand what you mean here, could you clarify?

File: flash/text/seq2seq/core/model.py (hunk 8 above)
Comment (SeanNaren) [reply flash/text/seq2seq/core/model.py:64]: You mean reset this back to False? I think it's fine for now, we can decide later (it's also an issue in text classification)

File: docs/source/reference/summarization.rst (hunk 3 above)
Comment (Borda) [reply docs/source/reference/summarization.rst:65]: see the alignment, it seems to. be that som section is tab (imports) and set of the code is 4 spaces

File: flash/text/seq2seq/core/model.py (hunk 8 above)
Comment (Borda) [reply flash/text/seq2seq/core/model.py:64]: then I would move it to upper for example to init

Comment (SeanNaren) [reply flash/text/seq2seq/core/model.py:64]: Ah but that will add clutter to the example, I think it's best the logic is resolved within a base class (that's how it will be handled eventually!). Definitely something important, but not too big an issue to deal with now I think!

File: flash_examples/predict/translate.py (hunk 11 above)
Comment (Borda) [reply flash_examples/predict/translate.py:6]: my point is if all code in examples is stated in `if __name__ == "__main__":` there is no reason to run tests here
and I would rather remove this intention so it is cleaner...

Comment (tchaton) [reply flash_examples/predict/translate.py:6]: It is there for num_workers > 1

File: docs/source/reference/text_classification.rst [hunk 13]
```diff
@@ -83,10 +83,10 @@ All we need is three lines of code to train our model!
 
 .. code-block:: python
```
Comment (Borda) [inline docs/source/reference/text_classification.rst:84]:
```suggestion
.. testcode::
```

File: docs/source/reference/translation.rst [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+.. note:: When changing the backbone, make sure you pass in the same backbone to the Task and the Data object! Since this is a Seq2Seq task, make sure you use a Seq2Seq model.
+
+.. code-block:: python
+
+    datamodule = TranslationData.from_files(
+        train_file="data/wmt_en_ro/train.csv",
+        valid_file="data/wmt_en_ro/valid.csv",
+        test_file="data/wmt_en_ro/test.csv",
+        input="input",
+        target="target",
```
Comment (Borda) [inline docs/source/reference/translation.rst:139]: shall this be default?

File: flash/text/seq2seq/core/model.py [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+
+    def compute_metrics(self, generated_tokens, batch, prefix):
+        pass
+
+    @property
+    def task(self) -> Optional[str]:
+        """
+        Override to define AutoConfig task specific parameters stored within the model.
+        """
+        return None
```
Comment (Borda) [inline flash/text/seq2seq/core/model.py:109]:
```suggestion
        pass
```

File: flash/text/seq2seq/core/model.py [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+    def _initialize_model_specific_parameters(self):
+        task_specific_params = self.model.config.task_specific_params
+
+        if task_specific_params is not None:
+            pars = task_specific_params.get(self.task, {})
+            rank_zero_info(f"Overriding model paramameters for {self.task} as defined within the model:\n {pars}")
+            self.model.config.update(pars)
+
+    @property
+    def tokenizer(self):
```
Comment (Borda) [inline flash/text/seq2seq/core/model.py:120]: typing? to all these methods...

File: flash/text/seq2seq/summarization/metric.py [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+        for key in self._defaults.keys():
+            value = getattr(self, key)
+            if isinstance(value, list):
+                value = tuple(value)
+            hash_vals.append(value)
+
+        return hash(tuple(hash_vals))
+
+
+class RougeBatchAggregator(scoring.BootstrapAggregator):
```
Comment (Borda) [inline flash/text/seq2seq/summarization/metric.py:74]: any docstring?

File: flash/text/seq2seq/summarization/utils.py [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import re
+
+from filelock import FileLock
+
+try:
+    import nltk
+
+    NLTK_AVAILABLE = True
```
Comment (Borda) [inline flash/text/seq2seq/summarization/utils.py:21]: rather use `_module_available` from PL

File: flash_examples/finetuning/translation.py [hunk 19]
```diff
@@ -0,0 +1,31 @@
+import flash
+from flash import download_data
+from flash.text import TranslationData, TranslationTask
+
+if __name__ == "__main__":
```
Comment (Borda) [inline flash_examples/finetuning/translation.py:5]: I am not much fun of this script when all body is hidden in if/main

File: flash_notebooks/generic_task.ipynb [hunk 20]
```diff
@@ -91,139 +83,34 @@
...
    "source": [
     "### 3. Split the data randomly"
    ]
   },
   {
    "cell_type": "code",
-   "execution_count": 12,
-   "id": "indonesian-arlington",
+   "execution_count": null,
+   "id": "southwest-muscle",
```
Comment (Borda) [inline flash_notebooks/generic_task.ipynb:113]: what is this change about?

**Review by Borda — COMMENTED** [review]
what is `flash_notebooks/.lock` for?
why the strange notebook changes?

File: flash_examples/finetuning/translation.py (hunk 19 above)
Comment (SeanNaren) [reply flash_examples/finetuning/translation.py:5]: this is primarily done for multiprocessing reasons, i think mp proc has issues when you do not include an if/main. If anything this is the format that the other scripts have been using!

File: docs/source/reference/translation.rst (hunk 14 above)
Comment (SeanNaren) [reply docs/source/reference/translation.rst:139]: target isn't default unfortunately, as this would break prediction support! can refactor in the future

**Review by tchaton — APPROVED** [review]
